ql/src/java/org/apache/hadoop/hive/ql/io/parquet/serde/ParquetHiveSerDe.java

```diff
@@ -163,4 +175,89 @@ public SerDeStats getSerDeStats() {
     }
     return stats;
   }
+
+  /**
+   * Given a complete struct type info and pruned paths containing selected fields
+   * from the type info, return a pruned struct type info only with the selected fields.
+   *
+   * For instance, if 'completeTypeInfo' is: s:struct<a:struct<b:int, c:boolean>, d:string>
+   *   and 'prunedPaths' is "s.a.b,s.d", then the result will be:
+   *   s:struct<a:struct<b:int>, d:string>
+   *
+   * @param completeTypeInfo the complete struct type info
+   * @param prunedPaths a string representing the pruned paths, separated by ','
+   * @return the pruned struct type info
+   */
+  private StructTypeInfo pruneFromPaths(
+      StructTypeInfo completeTypeInfo, String prunedPaths) {
+    PrunedTypeInfo prunedTypeInfo = new PrunedTypeInfo(completeTypeInfo);
+
+    String[] prunedPathList = prunedPaths.split(",");
+    for (String path : prunedPathList) {
+      pruneFromSinglePath(prunedTypeInfo, path);
+    }
+
+    return prunedTypeInfo.prune();
+  }
+
+  private void pruneFromSinglePath(PrunedTypeInfo prunedInfo, String path) {
+    Preconditions.checkArgument(prunedInfo != null,
+        "PrunedTypeInfo for path " + path + " should not be null");
+
+    int index = path.indexOf('.');
+    if (index < 0) {
+      index = path.length();
+    }
+
+    String fieldName = path.substring(0, index);
+    prunedInfo.markSelected(fieldName);
+    if (index < path.length()) {
+      pruneFromSinglePath(prunedInfo.children.get(fieldName), path.substring(index + 1));
+    }
+  }
+
+  private static class PrunedTypeInfo {
+    final StructTypeInfo typeInfo;
+    final Map<String, PrunedTypeInfo> children;
+    final boolean[] selected;
+
+    PrunedTypeInfo(StructTypeInfo typeInfo) {
+      this.typeInfo = typeInfo;
+      this.children = new HashMap<>();
+      this.selected = new boolean[typeInfo.getAllStructFieldTypeInfos().size()];
+      for (int i = 0; i < typeInfo.getAllStructFieldTypeInfos().size(); ++i) {
+        TypeInfo ti = typeInfo.getAllStructFieldTypeInfos().get(i);
+        if (ti.getCategory() == Category.STRUCT) {
+          this.children.put(typeInfo.getAllStructFieldNames().get(i),
+              new PrunedTypeInfo((StructTypeInfo) ti));
+        }
+      }
+    }
+
+    void markSelected(String fieldName) {
+      int index = typeInfo.getAllStructFieldNames().indexOf(fieldName);
+      if (index >= 0) {
+        selected[index] = true;
+      }
+    }
+
+    StructTypeInfo prune() {
+      List<String> newNames = new ArrayList<>();
+      List<TypeInfo> newTypes = new ArrayList<>();
+      List<String> oldNames = typeInfo.getAllStructFieldNames();
+      List<TypeInfo> oldTypes = typeInfo.getAllStructFieldTypeInfos();
+      for (int i = 0; i < oldNames.size(); ++i) {
+        String fn = oldNames.get(i);
+        if (selected[i]) {
+          newNames.add(fn);
+          if (children.containsKey(fn)) {
+            newTypes.add(children.get(fn).prune());
```

Comment:
Is this method be used as follows?
```
path: a.b.c
type: a:struct<b:struct<c:string,d:double>,e:string>
returns: a:struct<b:struct<c:string>>
```
If so, it may be problematic since we don't set the ```select``` field recursively for each leaf in L197. If I understand right, please add a test case to cover it as well.